src/Mapless-SQLite-Core/MaplessSQLitePool.class.st

```diff
@@ -0,0 +1,245 @@
+
+{ #category : #accessing }
+MaplessSQLitePool >> busyClients [
+	"Answers the bunch of clients that are currently busy"
+
+	^ busyClients ifNil: [ self initializeBusyClients ]
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> clientsMonitor [
+	"Answers the monitor used to protect clients usage"
+
+	^ clientsMonitor ifNil: [ self initializeClientsMonitor ]
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> currentClientDynamicVariable [
+	"Answers the class of the dynamic variable that will
+	hold the value of the current client during the evaluation of its usage."
+
+	^ MaplessSQLiteCurrentClient
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> databaseName [
+	^ databaseName
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> databaseName: anObject [
+	databaseName := anObject
+]
+
+{ #category : #testing }
+MaplessSQLitePool >> hasBusyClients [
+	"Answers true if this pool has, at least one, busy client"
+
+	^ self busyClients notEmpty
+]
+
+{ #category : #testing }
+MaplessSQLitePool >> hasIdleClients [
+	"Answers true if this pool has, at least one, idle client"
+
+	^ self idleClients notEmpty
+]
+
+{ #category : #testing }
+MaplessSQLitePool >> hasMaxClients [
+	^ self maxClients = self quantityOfClients
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> idleClients [
+	"Answers the bunch of clients that are currently idle"
+
+	^ idleClients ifNil: [ self initializeIdleClients ]
+]
+
+{ #category : #actions }
+MaplessSQLitePool >> ifUnavailablePurge: aMaplessDbClient [
+	"Tests aMaplessDbClient and remove it from this pool if invalid."
+
+	aMaplessDbClient isValid
+		ifFalse: [ self remove: aMaplessDbClient ]
+]
+
+{ #category : #initialization }
+MaplessSQLitePool >> initializeBusyClients [
+	^ busyClients := IdentitySet new
+]
+
+{ #category : #initialization }
+MaplessSQLitePool >> initializeClientsMonitor [
+	^ clientsMonitor := Monitor new
+]
+
+{ #category : #initialization }
+MaplessSQLitePool >> initializeIdleClients [
+	^ idleClients := IdentitySet new
+]
+
+{ #category : #initialization }
+MaplessSQLitePool >> initializeMaxClients [
+	^ maxClients := self class defaultMaxClients
+]
+
+{ #category : #initialization }
+MaplessSQLitePool >> initializeOn: aDatabaseFilenameOrKeyword [
+	databaseFilenameOrKeyword := aDatabaseFilenameOrKeyword
+]
+
+{ #category : #actions }
+MaplessSQLitePool >> makeClient [
+	"Returns a new client that can be (re)used in the pool"
+
+	^ MaplessSQLiteClient openOn: databaseFilenameOrKeyword
+]
+
+{ #category : #accessing }
+MaplessSQLitePool >> maxClients [
+	^ maxClients ifNil: [ self initializeMaxClients ]
+]
+
+{ #category : #reactions }
+MaplessSQLitePool >> onLogout: maplessDbClient [
+	self remove: maplessDbClient
+]
+
+{ #category : #actions }
+MaplessSQLitePool >> printOn: aStream [
+	super printOn: aStream.
+	aStream
+		nextPutAll:
+			'(' , self busyClients size asString , ' busy, '
+				, self idleClients size asString , ' idle @'
+				, databaseFilenameOrKeyword , ')'
```

Comment:
since you have a stream, I would rather avoid concatenation and use cascading `nextPutAll:` messages